tests/test_nhg.py

```diff
@@ -138,3 +139,187 @@ def test_route_nhg(self, dvs, testlog):
+        route_count = 0
+        r = 0
+        while route_count < MAX_ECMP_COUNT:
+            r += 1
+            fmt = '{{0:0{}b}}'.format(MAX_PORT_COUNT)
+            binary = fmt.format(r)
+            # We need at least 2 ports for a nexthop group
+            if binary.count('1') <= 1:
+                continue
+            fvs = gen_nhg_fvs(binary)
+            route_ipprefix = gen_ipprefix(route_count)
+            ps.set(route_ipprefix, fvs)
+            route_count += 1
+
+        asic_db = dvs.get_asic_db()
+        rtbl = swsscommon.Table(asic_db.db_connection, "ASIC_STATE:SAI_OBJECT_TYPE_ROUTE_ENTRY")
+        nhgtbl = swsscommon.Table(asic_db.db_connection, "ASIC_STATE:SAI_OBJECT_TYPE_NEXT_HOP_GROUP")
+
+        # Wait and check ASIC DB the count of nexthop groups used
+        asic_db.wait_for_n_keys("ASIC_STATE:SAI_OBJECT_TYPE_NEXT_HOP_GROUP", MAX_ECMP_COUNT)
+        asic_routes_count = len(rtbl.getKeys())
+
+        # Add second batch of routes with unique nexthop groups in AppDB
+        # Add more routes with new nexthop group in AppDBdd
+        route_ipprefix = gen_ipprefix(route_count)
+        base_ipprefix = route_ipprefix
+        base = route_count
+        route_count = 0
+        while route_count < 10:
+            r += 1
+            fmt = '{{0:0{}b}}'.format(MAX_PORT_COUNT)
+            binary = fmt.format(r)
+            # We need at least 2 ports for a nexthop group
+            if binary.count('1') <= 1:
+                continue
+            fvs = gen_nhg_fvs(binary)
+            route_ipprefix = gen_ipprefix(base + route_count)
+            ps.set(route_ipprefix, fvs)
+            route_count += 1
+        last_ipprefix = route_ipprefix
+
+        # Wait until we get expected routes and check ASIC DB on the count of nexthop groups used, and it should not increase
+        keys = asic_db.wait_for_n_keys("ASIC_STATE:SAI_OBJECT_TYPE_ROUTE_ENTRY", asic_routes_count + 10)
+        asic_db.wait_for_n_keys("ASIC_STATE:SAI_OBJECT_TYPE_NEXT_HOP_GROUP", MAX_ECMP_COUNT)
+
+        # Check the route points to next hop group
+        # Note: no need to wait here
+        k = asic_route_exists(keys, "2.2.2.0/24")
+        assert k is not None
+        fvs = asic_route_nhg_fvs(k)
+        assert fvs is not None
+
+        # Check the second batch does not point to next hop group
+        k = asic_route_exists(keys, base_ipprefix)
+        assert k is not None
+        fvs = asic_route_nhg_fvs(k)
+        assert fvs is None
+
+        # Remove first batch of routes with unique nexthop groups in AppDB
+        route_count = 0
+        r = 0
+        while route_count < MAX_ECMP_COUNT:
+            r += 1
+            fmt = '{{0:0{}b}}'.format(MAX_PORT_COUNT)
+            binary = fmt.format(r)
+            # We need at least 2 ports for a nexthop group
+            if binary.count('1') <= 1:
+                continue
+            route_ipprefix = gen_ipprefix(route_count)
+            ps._del(route_ipprefix)
```

Comment:
dvs database also has a `delete_entry` method you can use.

Reply:
ProducerStateTable is not supported, only Table.

---
In reply to: [420339845](https://github.com/Azure/sonic-swss/pull/1238#discussion_r420339845) [](ancestors = 420339845)